README.md
<!--
This file is written using Markdown language, which might make it difficult to
read it in a plain text editor. Please, visit ICON project page on DKRZ GitLab
(https://gitlab.dkrz.de/icon/icon) to see this file rendered or use a Markdown
viewer of your choice (https://www.google.com/search?q=markdown+viewer).
-->

# Introduction

The **ICON modelling framework** is a joint project between the
[German Weather Service](https://www.dwd.de/EN/Home/home_node.html) and the
[Max Planck Institute for Meteorology](https://www.mpimet.mpg.de/en/home/) for
developing a unified next-generation global numerical weather prediction and
climate modelling system. The ICON model has been introduced into DWD's
operational forecast system in January 2015.

This document provides technical details on how to build the model, as well as
the description of basic steps required to run it. More information about
ICON is available in the
[corresponding section](https://code.mpimet.mpg.de/projects/iconpublic/wiki/Documentation)
of the
[project's public web page](https://code.mpimet.mpg.de/projects/iconpublic).

# Table of contents
1. [Quick start](#quick-start)
2. [Configuration](#configuration)
    - [ICON dependencies](#icon-dependencies)
    - [Bundled libraries](#bundled-libraries)
    - [Compilers and tools](#compilers-and-tools)
    - [Compiler flags](#compiler-flags)
        - [Fortran compile groups](#fortran-compile-groups)
    - [Dynamic libraries](#dynamic-libraries)
    - [Configuration and building environments](#configuration-and-building-environments)
    - [Configuration wrappers](#configuration-wrappers)
    - [Out-of-source configuration (building)](#out-of-source-configuration-building)
3. [Building](#building)
    - [Source file collection](#source-file-collection)
    - [Preprocessing](#preprocessing)
    - [Source dependency tracking](#source-dependency-tracking)
        - [Code consistency checks](#code-consistency-checks)
        - [Compilation cascade prevention](#compilation-cascade-prevention)
    - [Building of the bundled libraries](#building-of-the-bundled-libraries)
    - [Source provenance collection](#source-provenance-collection)
4. [Running](#running)
    - [Out-of-build runs](#out-of-build-runs)
5. [FAQ](#faq)

# Quick start

The process of building ICON consists of two parts: *configuring* the options
and compiler flags, and *building* the source code with those options and flags.

The configuration step is normally done by running the
[./configure](./configure) script with command-line arguments, which, among
other things, tells the script where to locate libraries and tools required for
building. The list of arguments enabling a successful configuration might be
quite long and difficult to compose, therefore, instead of running the generic
[./configure](./configure) script directly, users are recommended to execute a
corresponding platform- or machine-specific *configuration wrapper* that sets
the required compiler and linker flags as well as the recommended set of
configure options. The wrappers can be found in the respective subdirectories of
the [./config](./config) directory.

For example, if you need to build ICON on
[Levante@DKRZ](https://docs.dkrz.de/doc/levante/index.html) with OpenMP feature
enabled using Intel compiler, you can run:

```console
$ cd /path/to/icon
$ ./config/dkrz/levante.intel --enable-openmp
```

Alternatively, you can create a directory and perform an *out-of-source* build:

```console
$ mkdir build && cd build
$ /path/to/icon/config/dkrz/levante.intel --enable-openmp
```

This way, you can build ICON in several different configurations, i.e. with
different compilers and features, using the same copy of the source code.

> **_NOTE:_** If you want to build and run ICON on your personal
computer/laptop, consider using [generic configure wrappers](./config/generic).

> **_NOTE:_** If there is no configure wrapper script for your platform or
machine, refer to section [Configuration](#configuration) for information on how
to work with the [./configure](./configure) script directly.

The building step is done by running `make` command with an optional argument
specifying the number of jobs to run simultaneously. For example,

```console
$ make -j8
```

The result of the building &mdash; the executable file of ICON &mdash; is saved
to the `./bin` subdirectory of the build directory.

> **_NOTE:_** For more information on the building step, refer to section
[Building](#building).

In case of an *in-source* build the *runscripts* can be generated by running:

```console
$ ./make_runscripts --all
```

In case of an *out-of-source* build the *runscripts* can be generated by running
the command above in the build direcotry. (A copy of `make_runscripts` should
have been placed there by any of the configuration wrappers. Also see section 
[Running](#running)..)

To run the model, switch to the `run` subdirectory of the build directory and
submit the runscript of your choice, for example:

```console
$ cd ./run && sbatch ./exp.atm_amip.run
```

> **_NOTE:_** For more information on the runscript generation and running the
model, refer to section [Running](#running).

# Configuration

The configuration step is done by calling the [configure](./configure) script
with arguments specifying the location of libraries and tools required for
building, as well as options enabling or disabling particular features of the
model. For example:

```console
$ ./configure CC=mpicc FC=mpif90 LIBS='-lnetcdff -lnetcdf -llapack -lblas' --disable-ocean --disable-coupling
```

> **_NOTE:_** Users are recommended to get familiar with the full list of
configuration options and their default values by running:
>```console
>$ ./configure --help
>```

The [configure](./configure) script of ICON is implemented using
[Autoconf](https://www.gnu.org/software/autoconf/) and its interface should be
familiar to those who have experience with
[Autotools](https://en.wikipedia.org/wiki/GNU_Autotools)-based building systems.

The following sections provide information on some features and implementation
details of the configuration process.

## ICON dependencies

[Fig. 1](#icon-depgraph) shows a partial dependency graph of the model. A
dependency can be either *mandatory* (i.e. the library is required regardless of
the specified configure options) or *optional* (i.e. the library is required
only if particular features of the model are enabled), and some of the
dependencies are provided together with the source code of ICON as git
submodules and refered to as *bundled* later in the text.

> **_NOTE:_** The term *bundled library* does not apply to all packages listed
in [.gitmodules](/.gitmodules): some of them, e.g. JSBACH and ART, have circular
dependencies with the ICON source code and therefore are treated as part of it.

<a name="icon-depgraph"/>![ICON dependency graph](./doc/config/icon-config-doc-depgraph.svg)

*Fig. 1. ICON dependency graph*

The list of libraries (packages) required for successful configuration and
building depends on the selected options. To make the configuration process more
transparent, the [configure](./configure) script does not accept paths to the
installation directories of the packages, which would be used to extend the
corresponding compiler flags. Instead, paths to the header and library files of
the packages must be provided by the user as compiler and linker flags, i.e. in
the `FCFLAGS`, `CPPFLAGS`, and `LDFLAGS` arguments. Moreover, the script does
not try to guess the list of libraries to be used, therefore all the `-l` linker
flags need to be specified in the `LIBS` argument in the correct order. The
recommended (topologically sorted) order for the `LIBS` argument is presented in
[Table 1](#icon-deptable), and the recommended order for the `FCFLAGS`,
`CPPFLAGS`, `LDFLAGS` is the reversed one.

<a name="icon-deptable">*Table 1. Topologically sorted list of the ICON dependency graph*</a>

| Linking order | Package | Dependency condition<sup><a name="f1-back" href="#f1">1</a></sup> | Required flags<sup><a href="#f1">1</a></sup> |
| :---: | :---: | :---: | :---: |
|  1 | [ICON-TIXI](https://gitlab.dkrz.de/icon-libraries/libtixi) (a modified version of [TIXI](https://github.com/DLR-SC/tixi)) | `--enable-art --with-external-tixi` | `FCFLAGS='-I/path/to/tixi/include' LDFLAGS='-L/path/to/tixi/lib' LIBS='-licon_tixi'` |
|  2 | [YAC](https://doc.redmine.dkrz.de/YAC/html/) | `--enable-coupling --with-external-yac` | `FCFLAGS='-I/path/to/yac/include' LDFLAGS='-L/path/to/yac/lib' LIBS='-lyac'` |
|  3 | [XML2](http://www.xmlsoft.org/) | `--enable-coupling`<sup><a name="f2-back" href="#f2">2</a></sup> or `--enable-art`<sup><a name="f2-back" href="#f2">2</a></sup>| `CPPFLAGS='-I/path/to/libxml2/include/libxml2' LDFLAGS='-L/path/to/libxml2/lib' LIBS='-lxml2'` |
|  4 | [MTIME](https://gitlab.dkrz.de/icon-libraries/libmtime) (Fortran interface) | `--with-external-mtime`<sup><a name="f3-back" href="#f3">3</a></sup> | `FCFLAGS='-I/path/to/mtime/include' LDFLAGS='-L/path/to/mtime/lib' LIBS='-lmtime'` |
|  5 | [MTIME](https://gitlab.dkrz.de/icon-libraries/libmtime) (C interface) | `--enable-coupling --without-external-yac --with-external-mtime` | `CPPFLAGS='-I/path/to/mtime/include' LDFLAGS='-L/path/to/mtime/lib' LIBS='-lmtime'` |
|  6 | [SERIALBOX](https://github.com/GridTools/serialbox) | `--enable-serialization` | `FCFLAGS='-I/path/to/serialbox2/include' LDFLAGS='-L/path/to/serialbox2/lib' LIBS='-lSerialboxFortran'` |
|  7 | [CDI](https://code.mpimet.mpg.de/projects/cdi/) | `--with-external-cdi` | `FCFLAGS='-I/path/to/libcdi/include' LDFLAGS='-L/path/to/libcdi/lib' LIBS='-lcdi_f2003 -lcdi'` (or `LIBS='-lcdi_f2003 -lcdipio -lcdi'`) |
|  8 | [PPM](https://gitlab.dkrz.de/jahns/ppm) (C interface) | `--enable-cdi-pio --with-external-ppm --without-external-cdi` |  `CPPFLAGS='-I/path/to/ppm/include' LDFLAGS='-L/path/to/ppm/lib' LIBS='-lscalesppmcore'` |
|  9 | [ECCODES](https://confluence.ecmwf.int/display/ECC) (Fortran interface) | `--enable-emvorado` | `FCFLAGS='-I/path/to/eccodes/include' LDFLAGS='-L/path/to/eccodes/lib' LIBS='-leccodes_f90'` |
| 10 | [ECCODES](https://confluence.ecmwf.int/display/ECC) or [GRIB-API](https://www.ecmwf.int/en/newsletter/152/news/end-road-grib-api)<sup><a name="f4-back" href="#f4">4</a></sup> (C interface) | `--enable-grib2 --without-external-cdi` | `CPPFLAGS='-I/path/to/eccodes/include' LDFLAGS='-L/path/to/eccodes/lib' LIBS='-leccodes'` (or `LIBS='-lgrib_api'`) |
| 11 | [YAXT](https://gitlab.dkrz.de/dkrz-sw/yaxt) (Fortran interface) | `--enable-yaxt --with-external-yaxt` or `--enable-cdi-pio --with-external-yaxt`<sup><a name="f5-back" href="#f5">5</a></sup> | `FCFLAGS='-I/path/to/yaxt/include' LDFLAGS='-L/path/to/yaxt/lib' LIBS='-lyaxt'` |
| 12 | [YAXT](https://gitlab.dkrz.de/dkrz-sw/yaxt) (C interface) | `--enable-cdi-pio --without-external-cdi --with-external-yaxt` or `--enable-coupling --with-external-yaxt`<sup><a name="f6-back" href="#f6">6</a>, <a name="f7-back" href="#f7">7</a></sup> | `CPPFLAGS='-I/path/to/yaxt/include' LDFLAGS='-L/path/to/yaxt/lib' LIBS='-lyaxt_c'` |
| 13 | [SCT](https://gitlab.dkrz.de/dkrz-sw/sct) (Fortran interface) | `--enable-sct --with-external-sct` | `FCFLAGS='-I/path/to/sct/include' LDFLAGS='-L/path/to/sct/lib' LIBS='-lsct'` |
| 14 | [RTTOV](https://www.nwpsaf.eu/site/software/rttov/) | `--enable-rttov` | `FCFLAGS='-I/path/to/rttov/include -I/path/to/rttov/mod' LDFLAGS='-L/path/to/rttov/lib' LIBS='-lrttov_other -lrttov_emis_atlas -lrttov_brdf_atlas -lrttov_parallel -lrttov_coef_io -lrttov_hdf -lrttov_main'` |
| 15 | [LAPACK](http://www.netlib.org/lapack/) (or analogue) | mandatory | `LDFLAGS='-L/path/to/lapack/lib' LIBS='-llapack'` (depends on the implementation) |
| 16 | [BLAS](http://www.netlib.org/blas/) (or analogue) | mandatory | `LDFLAGS='-L/path/to/blas/lib' LIBS='-lblas'` (depends on the implementation) |
| 17 | [ECRAD](https://confluence.ecmwf.int/display/ECRAD/ECMWF+Radiation+Scheme+Home) | `--enable-ecrad --with-external-ecrad` | `FCFLAGS='-I/path/to/ecrad/include' LDFLAGS='-L/path/to/ecrad/lib' LIBS='-lradiation -lifsrrtm -lutilities -lifsaux'` |
| 18 | [RTE+RRTMGP](https://github.com/earth-system-radiation/rte-rrtmgp) | `--enable-rte-rrtmgp --with-external-rte-rrtmgp` | `FCFLAGS='-I/path/to/rte-rrtmgp/include' LDFLAGS='-L/path/to/rte-rrtmgp/lib' LIBS='-lrrtmgp -lrte'` |
| 19 | [NetCDF-Fortran](https://www.unidata.ucar.edu/software/netcdf/docs-fortran/) | mandatory | `FCFLAGS='-I/path/to/netcdf-fortran/include' LDFLAGS='-L/path/to/netcdf-fortran/lib' LIBS='-lnetcdff'` |
| 20 | [NetCDF-C](https://www.unidata.ucar.edu/software/netcdf/docs/) | `--without-external-cdi` or `--enable-coupling`<sup><a name="f8-back" href="#f8">8</a></sup> | `CPPFLAGS='-I/path/to/netcdf/include' LDFLAGS='-L/path/to/netcdf/lib' LIBS='-lnetcdf'` |
| 21 | [HDF5](https://support.hdfgroup.org/HDF5/) (low- and high-level Fortran interfaces) | `--enable-emvorado` or `--enable-rttov`<sup><a name="f9-back" href="#f9">9</a></sup> | `FCFLAGS='-I/path/to/hdf5/include' LDFLAGS='-L/path/to/hdf5/lib' LIBS='-lhdf5_hl_fortran -lhdf5_fortran'` |
| 22 | [HDF5](https://support.hdfgroup.org/HDF5/) (low-level C interface) | `--enable-sct --without-external-sct` | `CPPFLAGS='-I/path/to/hdf5/include' LDFLAGS='-L/path/to/hdf5/lib' LIBS='-lhdf5'` |
| 23 | [ZLIB](https://zlib.net/) | `--enable-emvorado` | `LDFLAGS='-L/path/to/zlib/lib' LIBS='-lz'`<sup><a name="f10-back" href="#f10">10</a></sup> |
| 24 | [AEC](https://gitlab.dkrz.de/k202009/libaec) or [SZIP](https://support.hdfgroup.org/doc_resource/SZIP/) | static linking | `LDFLAGS='-L/path/to/aec/lib' LIBS='-laec'` (or `LIBS='-lsz'`) |
| 25 | [MPI](https://www.mpi-forum.org/) (Fortran interface) | `--enable-mpi`<sup><a name="f11-back" href="#f11">11</a></sup> or `--enable-yaxt` or `--enable-cdi-pio --without-external-cdi --without-external-yaxt` or `--enable-coupling --without-external-yac` | `FC='/path/to/mpi/bin/mpif90'` or `FCFLAGS='-I/path/to/mpi/include' LDFLAGS='-L/path/to/mpi/lib' LIBS='-lmpifort -lmpi'` (depends on the implementation) |
| 26 | [MPI](https://www.mpi-forum.org/) (C interface) | `--enable-coupling`<sup><a name="f12-back" href="#f12">12</a></sup> or `--enable-yaxt --without-external-yaxt` or `--enable-mpi --enable-sct --without-external-sct` or `--enable-cdi-pio --without-external-cdi` | `CC=/path/to/mpi/bin/mpicc` or `CPPFLAGS=-I/path/to/mpi/include LDFLAGS='-L/path/to/mpi/lib' LIBS='-lmpi'` (depends on the implementation) |
| 27 | [CUB](https://nvlabs.github.io/cub/) | `--enable-gpu --with-external-cub` | `NVCFLAGS=-I/path/to/cub`<sup><a name="f13-back" href="#f13">13</a></sup> |
| 28 | [CUDA](https://developer.nvidia.com/cuda-zone) | `--enable-gpu` | `LDFLAGS=-L/path/to/cuda/lib LIBS=-lcudart`<sup><a name="f14-back" href="#f14">14</a></sup> |
| 29 | [STDC++](https://gcc.gnu.org/onlinedocs/libstdc++/) | `--enable-gpu` | `LDFLAGS=-L/path/to/gcc/used/by/nvcc/lib LIBS=-lstdc++` |

1. <a name="f1"/> The dependency conditions and required flags are specified
assuming that the shared versions of the libraries containing `RPATH` entries
pointing to their dependencies are used. Also note that the configure scripts of
some of the bundled libraries (e.g. YAXT) run checks that link executables using
libtool and then run them. Since some of the dependencies (e.g. ECCODES) are not
libtool libraries (i.e. there are not .la files for them), the resulting
executables do not have the corresponding `RPATH` entries, which triggers false
negative results of the checks and fails the configuration with misleading
messages. To circumvent this problem, the paths to the non-libtool dependencies
must be passed to the linker before running the configure script of ICON, e.g.
by means of the `LD_LIBRARY_PATH` environment variable. [↩](#f1-back)
2. <a name="f2"/> There are no shared versions of YAC and ICON-TIXI libraries,
which could link XML2 library implicitly, therefore, the latter needs to be
linked explicitly, regardless of whether external or the bundled versions of YAC
and ICON-TIXI are used. [↩](#f2-back)
3. <a name="f3"/> When the coupling is enabled (`--enable-coupling`) and an
external version of YAC (`--with-external-yac`) is used, usage of an external
MTIME library (`--with-external-mtime`) is mandatory (must be the library that
YAC has been built with). [↩](#f3-back)
4. <a name="f4"/> GRIB-API is supported only in cases when ECCODES is not
required, e.g. when the radar forward operator EMVORADO, which requires the
Fortran interface of ECCODES, is disabled (`--disable-emvorado`). [↩](#f4-back)
5. <a name="f5"/> When usage of the parallel features of CDI is enabled
(`--enable-cdi-pio`) and an external version of CDI (`--with-external-cdi`) is
used, usage of an external YAXT library (`--with-external-yaxt`) is mandatory
(must be the library that CDI has been built with). [↩](#f6-back)
6. <a name="f6"/> When the coupling is enabled (`--enable-coupling`) and an
external version of YAC (`--with-external-yac`) is used, usage of an external
YAXT library (`--with-external-yaxt`) is mandatory (must be the library that YAC
has been built with). [↩](#f6-back)
7. <a name="f7"/> There is no shared version of YAC library, which could link
YAXT (C interface) library implicitly, therefore, the latter needs to be linked
explicitly, regardless of whether an external or the bundled version of YAC is
used. [↩](#f7-back)
8. <a name="f8"/> There is no shared version of YAC library, which could link
NetCDF-C library implicitly, therefore, the latter needs to be linked
explicitly, regardless of whether an external or the bundled version of YAC is
used. [↩](#f8-back)
9. <a name="f9"/> There is no shared version of RTTOV library, which could link
Fortran HDF5 libraries implicitly, therefore, the latter need to be linked
explicitly when usage of the radiative transfer model for TOVS is enabled
(`--enable-rttov`). [↩](#f9-back)
10. <a name="f10"/> ZLIB is used via the `ISO_C_BINDING` interface and does not
require additional preprocessor flags. [↩](#f10-back)
11. <a name="f11"/> When usage of the parallel features of CDI is enabled
(`--enable-cdi-pio`), MPI (parallelization) support (`--enable-mpi`) is
mandatory. [↩](#f11-back)
12. <a name="f12"/> There is no shared version of YAC library, which could link
MPI (C interface) library implicitly, therefore, the latter needs to be linked
explicitly, regardless of whether an external or the bundled version of YAC is
used. [↩](#f12-back)
13. <a name="f13"/> CUB is a header-only library and does not require additional
linker flags. [↩](#f13-back)
14. <a name="f14"/> Currently, the only Fortran compiler that supports all
features required to build the GPU version of ICON is PGI, which specifies
linker flags enabling CUDA automatically. [↩](#f14-back)

## Bundled libraries

As it was mentioned in the previous section, some of the libraries are bundled
together with the ICON source code. However, users can download and install
those libraries before configuring ICON and use them instead. This is controlled
by the `--with-external-<package>` arguments of the [configure](./configure)
script of ICON. The arguments accept either `yes` or `no`. If the usage of an
external version of a library is requested (i.e.
`--with-external-<package>=yes`), the compiler and linker flags, i.e. `FCFLAGS`,
`CPPFLAGS`, `LDFLAGS` and `LIBS`, are supposed to be extended accordingly (see
[Table 1](#icon-deptable)). The [configure](./configure) script of ICON fails if
the flags are not set correctly or enable a version of the library that is known
to be incompatible with ICON.


RELEASE_NOTES.md
# Release notes for icon-2.6.6

Later than expected: this is another ICON release

- OpenACC port for Nvidia GPUs and enabling HPE-Cray OpenACC use with AMD GPUs
  - almost all of the ICON atmospheric components are ready for use on GPU equiped machines. This has been a big effort by Nvidia, HPE/Cray, MeteoSwiss, DWD, CSCS, CSC, and MPIM (the order is arbitrary).
  - there is further progress coming up
- CSCS
  - preparation of daint/dom transition for Nvidia compiler bumb to 22.5.
  - added Docker files for EXCLAIM project in scripts/docker/exclaim
  - more test cases
- DWD
  - still persisting problem with OpenMP (non-reproducability) in buildbot tests.
  - bufixes and improvements in the data assimilation part and physics
  - two-moment cloud microphysics has been replaced by a new version
  - initial steps for wave model implementation
  - added jsbach/vdiff from former ICON echam-physics
  - enable/fix more problems with respect to run icon-seamless 
  - some code moves and consolidation
  - OpenACC beautification application running to allow passing a required CI test
  - more test cases
- MPIM
  - bugfixes in the ocean
  - added additional diagnostics
  - improvements in JSBACH (HD mask, LCC, former dynveg)
  - updating of build on levante and macos
  - fixes for claw
  - fixes for building ICON with some data assimilation components
  - Nvidia 22.9 is needed for JSBACH
  - ocean updated to zstar vertical coordinate
  - bump rte-rrtmgp to new version
- DKRZ
  - fix many errors and enabled new features with respect to coupling external components
  - new and updated config-wrappers
  - improve buildbot testing

For details take a look at:

https://gitlab.dkrz.de/icon/wiki/-/wikis/Protocol-of-Release-Commits

# Release notes for icon-2.6.5

It is time for yet another ICON release.

After 10 month we managed to provide a new ICON release with new
machine configurations, a new buildbot, significantly extended,
improved GPU capabilities, first steps into ICON seamless and a
cleaned up version of the MPIM AES physics (now available as aes
physics and not echam anymore.

- CSCS: Introduction of more and optimized OpenACC code parts and
        introducing work to allow later more modularization for
        improving and consolidating of the ICON code.

- DKRZ: Bugfixes and more bugfixes, cdi-1.8.4 incorporates nom (joint
        effort of DKRZ and CIMD) cdi-pio, new versions of all tools
        maintained by DKRZ, a new buildbot (Ralf Müller) and strong
        support for getting ICON running on levante.

- DWD: Introduction of many improvements in the parameterizations, in
       tuning and adding bugfixes. Setting up on the new ECMWF system
       and supporting the data assimilation system. First steps for
       ICON-seamless are done, especially the coupling to ICON ocean
       is technically available and most of the time varying external
       data for climate simulations have been added. However, the code
       variant has to be tuned and extensively tested and some parts are
       still missing. This activity is a joint effort of DWD and MPIM
       colleagues.

- MCH: Is actively providing more, and more model components of the
       NWP physics for the use on GPUs. Furthermore they are
       supporting all activities related to GPUs and testing of large
       parts of the ICON code.

- MPIM AES: To simplify and streamline further high resolution
       modelling only, the echam physics has been renamed as aes
       physics and some former components have been removed as are the
       SSO gravity drag parameterization, as well the gravity drag
       parameterization for the middle atmosphere, and most prominent
       the convection parameterization. As cloud paramterizations the
       DWD graupel and two-moment scheme are available. A few more,
       minor, changes have been applied. Newly added is the former
       large eddy model component the Smagorinsky 'vertical' diffusion
       scheme.

- MPIM OES: Added a new C-grid seaice model (C. Mehlmann) and extensive
       improvements of OpenMP for hamocc. 

- MPIM LES: JSBACH got more model components and has been adapted for
       to all changes above its top-level.  - KIT: The aerosoly
       microphysics and plume components have been added as well as a
       large number of bugfixes.

- MPIM CIMD: just a lot of things in the area of coupling, configuring
       and building and model infrastrucuture, new machine setups
       for lumi (CSC), juwels/booster (JSC) and levante (DKRZ) 

- All developers: many of the work has been done in cross
  institutional working groups.


A special thanks is going to Sergey Kosukhin (MPIM-CIMD) for his work
on the configuration and build system of ICON making life of
developers much easier.

The work on the GPU code version, implemented based on OpenACC has
been supported by NVIDIA. Thanks a lot, Dmitry Alexeev (and his
colleagues).

For detailed information, please, have a look at

https://gitlab.dkrz.de/icon/wiki/-/wikis/Protocol-of-Release-Commits

# Release notes for icon-2.6.4

Another new  release of icon is available.

It consists of

- fix for the soil water budget and snow-tiles
- integrate YAC2, corresponding changes in jsbach and the yaxt version are included
- bundled libraries are no checked at DWD and DKRZ

- fix current CDI-PIO setup for coupled setups
- use plain netcdf intead of CDI for restart IO

- fixes for the building system regarding new compilers, CLAW compatibility, changes in RTTOV and building against eccodes/grib_api

- bugfix for hfbasin diagnostic

- changes for the DWD NEC Aurora

- adjustments for ruby-0 setup (clean up of lsm masks, coupling, add irad_aero=12, buildbot test, scripts)

- New ICON-Land/JSBACH version: jsbach:master@39758f03
  This merge contains many changes, bug fixes, and improvements/new features in ICON-Land/JSBACH.
  The most important new features are:

src/shared/mo_var.f90
!! @par Copyright and License
!!
!! This code is subject to the DWD and MPI-M-Software-License-Agreement in
!! its most recent form.
!! Please see the file LICENSE in the root of the source tree for this code.
!! Where software is supplied by third parties, it is indicated in the
!! headers of the routines.

MODULE mo_var

  USE mo_kind,               ONLY: dp, sp
  USE mo_var_metadata_types, ONLY: t_var_metadata, t_var_metadata_dynamic, &
                                   VINTP_TYPE_LIST
  USE mo_exception,          ONLY: message, finish, message_text
  USE mo_impl_constants,     ONLY: REAL_T, SINGLE_T, BOOL_T, INT_T, &
    &                              STR_HINTP_TYPE
  USE mo_var_groups,         ONLY: var_groups_dyn

  IMPLICIT NONE
  PRIVATE

  ! export index constants: model/pressure/height levels
  PUBLIC :: level_type_ml, level_type_pl, level_type_hl, level_type_il, lev_type_str
  PUBLIC :: t_var, t_var_ptr

  ! constants defining level type:
  INTEGER, PARAMETER             :: level_type_ml = 1
  INTEGER, PARAMETER             :: level_type_pl = 2
  INTEGER, PARAMETER             :: level_type_hl = 3
  INTEGER, PARAMETER             :: level_type_il = 4
  ! string defining level type:
  CHARACTER(LEN=2), PARAMETER    :: lev_type_str(4) = (/ 'ML', 'PL', 'HL', 'IL' /)

  TYPE t_var
    REAL(dp), CONTIGUOUS, POINTER :: r_ptr(:,:,:,:,:) => NULL()
    REAL(sp), CONTIGUOUS, POINTER :: s_ptr(:,:,:,:,:) => NULL()
    INTEGER,  CONTIGUOUS, POINTER :: i_ptr(:,:,:,:,:) => NULL()
    LOGICAL,  CONTIGUOUS, POINTER :: l_ptr(:,:,:,:,:) => NULL()
    INTEGER                      :: var_base_size      ! generic size in bytes of variable used
    TYPE(t_var_metadata)         :: info               ! meta data for this entry
    TYPE(t_var_metadata_dynamic) :: info_dyn           ! dynamic meta data for this entry (see type description)
    TYPE(t_var), POINTER :: ref_to => NULL()
  CONTAINS
    PROCEDURE :: print_short => print_var_short
    PROCEDURE :: print_rigorous => print_var_rigorous
  END TYPE t_var

  TYPE t_var_ptr
    TYPE(t_var), POINTER :: p => NULL()
  END TYPE t_var_ptr

  CHARACTER(*), PARAMETER :: modname = "mo_var"

CONTAINS

  SUBROUTINE print_var_short(var)
    CLASS(t_var), INTENT(IN), TARGET :: var
    CHARACTER(LEN=1), PARAMETER :: lm3(3) = ['i', 'm', 'a'], &
                                   lm4(3) = ['c', 'v', 'e']
    CHARACTER(LEN=4) :: localMode
    TYPE(t_var_metadata), POINTER :: info

    info => var%info
    localMode = '----'
    IF (info%lrestart) localMode(1:1) = 'r'
    IF (info%lcontained) localMode(2:2) = 't'
    SELECT CASE (info%isteptype)
    CASE (1,2,3)
      localMode(3:3) = lm3(info%isteptype)
    END SELECT
    SELECT CASE (info%hgrid)
    CASE (1,2,3)
      localMode(4:4) = lm4(info%hgrid)
    CASE (45)
      localMode(4:4) = 'L'
    END SELECT
    WRITE(message_text, '(a4,3i4,a24,a40)') localMode, &
      & info%grib2%discipline, info%grib2%category,   &
      & info%grib2%number, TRIM(info%name), TRIM(info%cf%standard_name)
    CALL message('', message_text)
  END SUBROUTINE print_var_short

  SUBROUTINE print_var_rigorous(var)
    CLASS(t_var), INTENT(IN), TARGET :: var
    TYPE(t_var_metadata), POINTER :: info
    TYPE(t_var_metadata_dynamic), POINTER :: info_dyn
    CHARACTER(*), PARAMETER :: routine = modname//":print_var_rigorous", &
      & category(22) = [ &
      & 'Table entry name                            : ', &
      & 'Pointer status                              : ', &
      & 'Local field dimensions                      : ', &
      & 'Assigned GRIB discipline/category/parameter : ', &
      & 'CF convention standard name/unit            : ', &
      & 'CF convention long name                     : ', &
      & 'Field is in a container                     : ', &
      & ' Index in container                         : ', &
      & ' horizontal grid type used (C=1,V=2,E=3)    : ', &
      & ' vertical grid type used (see cdilib.c)     : ', &
      & ' stat. processing type (I=1,AVG=2,ACC=3...) : ', &
      & 'Missing value                               : ', &
      & 'Added to restart                            : ', &
      & 'Tracer field                                : ', &
      & '   Child-to-parent feedback                 : ', &
      & '   Horizontal transport method              : ', &
      & '   Vertical transport method                : ', &
      & '   Turbulent transport                      : ', &
      & 'Variable class/species                      : ', &
      & 'Variable group(s)                           : ', &
      & 'Horizontal interpolation                    : ', &
      & 'Vertical interpolation                      : ' ]
    CHARACTER(LEN=2) :: istr
    CHARACTER(LEN=20) :: mvstr
    CHARACTER(LEN=9) :: ipstr
    CHARACTER(LEN=32) :: dimstr
    INTEGER :: i, mlen, gnlen

    info => var%info
    info_dyn => var%info_dyn
    CALL message('', category(1)//TRIM(info%name))
    IF (ANY([ASSOCIATED(var%r_ptr), ASSOCIATED(var%s_ptr), &
             ASSOCIATED(var%i_ptr), ASSOCIATED(var%l_ptr)])) THEN
      CALL message ('',category(2)//'in use.')
      dimstr = '('
      DO i = 1, info%ndims
        WRITE(dimstr,"(a,i0,a1)") TRIM(dimstr), info%used_dimensions(i), &
          & MERGE(')', ',', info%ndims .EQ. i)
      ENDDO
      CALL message('', category(3)//TRIM(dimstr))
    ELSE
      CALL message('', category(2)//'not in use.')
    ENDIF
    WRITE (message_text,'(a,3i4)') category(4), &
      & info%grib2%discipline, info%grib2%category, info%grib2%number
    CALL message('', message_text)
    WRITE (message_text,'(a,a,a,a)') category(5), &
         TRIM(info%cf%standard_name), '     ', TRIM(info%cf%units)
    CALL message('', message_text)
    CALL message('', category(6)//TRIM(info%cf%long_name))
    CALL message('', category(7)//MERGE('yes.', 'no. ', info%lcontained))
    istr = '--'
    IF (info%lcontained) WRITE(istr, "(i2)") info%ncontained
    CALL message('', category(8)//istr)
    WRITE(istr, "(i2)") info%hgrid
    CALL message('', category(9)//istr)
    WRITE(istr, "(i2)") info%vgrid
    CALL message('', category(10)//istr)
    WRITE(istr, "(i2)") info%isteptype
    CALL message('', category(11)//istr)
    IF (info%lmiss) THEN
      SELECT CASE(info%data_type)
      CASE (REAL_T)
        WRITE(mvstr, "(e20.12)") info%missval%rval
      CASE (SINGLE_T)
        WRITE(mvstr, "(e20.12)") info%missval%sval
      CASE (INT_T)
        WRITE(mvstr, "(i8)") info%missval%ival
      CASE (BOOL_T)
        WRITE(mvstr, "(l8)") info%missval%lval
      END SELECT
    ELSE
      WRITE(mvstr, "(a)") 'off.'
    ENDIF
    CALL message('', category(12)//mvstr)
    CALL message('', category(13)//MERGE("yes.", " no.", info%lrestart))
    CALL message('', category(14)//MERGE("yes.", " no.", info_dyn%tracer%lis_tracer))
    IF (info_dyn%tracer%lis_tracer) THEN
      CALL message('', category(15)//MERGE("yes.", " no.", info_dyn%tracer%lfeedback))
      WRITE(ipstr, "(3i3)") info_dyn%tracer%ihadv_tracer
      CALL message('', category(16)//ipstr)
      WRITE(ipstr, "(3i3)") info_dyn%tracer%ivadv_tracer
      CALL message('', category(17)//ipstr)
      CALL message('', category(18)//MERGE("yes.", " no.", info_dyn%tracer%lturb_tracer))
    ENDIF !lis_tracer
    WRITE(istr, "(i2)") info%var_class
    CALL message('', category(19)//istr)
    IF (ANY(info%in_group(:))) THEN
      CALL message('', category(20))
      message_text = ' + '
      mlen = 3
      DO i = 1, var_groups_dyn%get_n_grps()
        IF (.NOT.info%in_group(i)) CYCLE
        gnlen = var_groups_dyn%gname_len(i)
        IF (mlen + 2 + gnlen .GT. LEN(message_text)) CALL finish(routine, "message_text overflow")
        WRITE(message_text, "(3a)") message_text(1:mlen), var_groups_dyn%gname(i)(1:gnlen), ", "
        mlen = mlen + 2 + gnlen
      END DO
      CALL message('', message_text(1:mlen-2))
    END IF
    CALL message('', category(21)//STR_HINTP_TYPE(info%hor_interp%hor_intp_type))
    DO i = 1, SIZE(VINTP_TYPE_LIST)
      IF (info%vert_interp%vert_intp_type(i)) &
        CALL message('', category(22)//VINTP_TYPE_LIST(i))
    END DO
    CALL message('', '')
  END SUBROUTINE print_var_rigorous

END MODULE mo_var
